File: java/FakeNewsModelo.java
import java.io.Serializable;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public final class FakeNewsModelo {

    public static final int IGNORANTE = 0;
    public static final int ESPALHADOR = 1;
    public static final int INATIVO = 2;

    private FakeNewsModelo() {
    }

    public static final class SimulationConfig implements Serializable {
        public int linhas = 500;
        public int colunas = 500;
        public int geracoes = 50;
        public double percentual = 0.02;
        public int limiar = 2;
        public long semente = 42L;
        public int threads = 4;

        public static SimulationConfig fromArgs(String[] args) {
            SimulationConfig config = new SimulationConfig();

            for (int i = 0; i < args.length; i++) {
                String argumento = args[i];
                if (!argumento.startsWith("--")) {
                    throw new IllegalArgumentException("Parametro invalido: " + argumento);
                }
                if (i + 1 >= args.length) {
                    throw new IllegalArgumentException("Valor ausente para: " + argumento);
                }

                String valor = args[++i];

                switch (argumento) {
                    case "--linhas":
                        config.linhas = Integer.parseInt(valor);
                        break;
                    case "--colunas":
                        config.colunas = Integer.parseInt(valor);
                        break;
                    case "--geracoes":
                        config.geracoes = Integer.parseInt(valor);
                        break;
                    case "--percentual":
                        config.percentual = Double.parseDouble(valor);
                        break;
                    case "--limiar":
                        config.limiar = Integer.parseInt(valor);
                        break;
                    case "--semente":
                        config.semente = Long.parseLong(valor);
                        break;
                    case "--threads":
                        config.threads = Integer.parseInt(valor);
                        break;
                    default:
                        throw new IllegalArgumentException("Parametro desconhecido: " + argumento);
                }
            }

            return config;
        }
    }

    public static final class StateCount {
        public int ignorantes;
        public int espalhadores;
        public int inativos;
    }

    public static int[][] createGrid(int linhas, int colunas, double percentualEspalhadores, long semente) {
        int[][] grade = createEmptyGrid(linhas, colunas);
        int totalCelulas = linhas * colunas;
        int totalEspalhadores = (int) (totalCelulas * percentualEspalhadores);

        if (totalEspalhadores <= 0) {
            return grade;
        }

        java.util.Random random = new java.util.Random(semente);
        Set<Integer> posicoes = new HashSet<>();

        while (posicoes.size() < totalEspalhadores) {
            posicoes.add(random.nextInt(totalCelulas));
        }

        for (int posicao : posicoes) {
            int linha = posicao / colunas;
            int coluna = posicao % colunas;
            grade[linha][coluna] = ESPALHADOR;
        }

        return grade;
    }

    public static int[][] createEmptyGrid(int linhas, int colunas) {
        return new int[linhas][colunas];
    }

    public static int[][] copyGrid(int[][] grade) {
        int[][] copia = new int[grade.length][grade[0].length];
        for (int i = 0; i < grade.length; i++) {
            copia[i] = Arrays.copyOf(grade[i], grade[i].length);
        }
        return copia;
    }

    public static int countSpreadersNeighbors(int[][] grade, int i, int j) {
        int linhas = grade.length;
        int colunas = grade[0].length;
        int total = 0;

        for (int di = -1; di <= 1; di++) {
            for (int dj = -1; dj <= 1; dj++) {
                if (di == 0 && dj == 0) {
                    continue;
                }

                int ni = i + di;
                int nj = j + dj;

                if (ni >= 0 && ni < linhas && nj >= 0 && nj < colunas && grade[ni][nj] == ESPALHADOR) {
                    total++;
                }
            }
        }

        return total;
    }

    public static int[] calculateNextRow(int[][] grade, int linha, int limiarConvencimento) {
        int[] novaLinha = Arrays.copyOf(grade[linha], grade[linha].length);

        for (int coluna = 0; coluna < grade[0].length; coluna++) {
            int estadoAtual = grade[linha][coluna];

            if (estadoAtual == IGNORANTE) {
                int vizinhos = countSpreadersNeighbors(grade, linha, coluna);
                novaLinha[coluna] = vizinhos >= limiarConvencimento ? ESPALHADOR : IGNORANTE;
            } else if (estadoAtual == ESPALHADOR) {
                novaLinha[coluna] = INATIVO;
            } else {
                novaLinha[coluna] = INATIVO;
            }
        }

        return novaLinha;
    }

    public static int[][] nextGenerationSequential(int[][] grade, int limiarConvencimento) {
        int[][] novaGrade = createEmptyGrid(grade.length, grade[0].length);
        for (int linha = 0; linha < grade.length; linha++) {
            novaGrade[linha] = calculateNextRow(grade, linha, limiarConvencimento);
        }
        return novaGrade;
    }

    public static StateCount countStates(int[][] grade) {
        StateCount contagem = new StateCount();

        for (int[] linha : grade) {
            for (int celula : linha) {
                if (celula == IGNORANTE) {
                    contagem.ignorantes++;
                } else if (celula == ESPALHADOR) {
                    contagem.espalhadores++;
                } else {
                    contagem.inativos++;
                }
            }
        }

        return contagem;
    }

    public static int[][] splitRanges(int totalLinhas, int quantidadePartes) {
        if (totalLinhas <= 0) {
            return new int[0][0];
        }

        quantidadePartes = Math.max(1, Math.min(quantidadePartes, totalLinhas));
        int base = totalLinhas / quantidadePartes;
        int resto = totalLinhas % quantidadePartes;
        int[][] faixas = new int[quantidadePartes][2];

        int inicio = 0;
        for (int indice = 0; indice < quantidadePartes; indice++) {
            int tamanho = base + (indice < resto ? 1 : 0);
            int fim = inicio + tamanho;
            faixas[indice][0] = inicio;
            faixas[indice][1] = fim;
            inicio = fim;
        }

        return faixas;
    }
}

File: python/FakeNewsModelo.py
"""Modelo compartilhado da simulacao de fake news.

Este modulo reune a logica de regra e manipulacao da grade.
Nao tem dependencia de threads ou sockets.
"""

from __future__ import annotations

import random

IGNORANTE = 0
ESPALHADOR = 1
INATIVO = 2


def copiar_grade(grade):
    # Cria uma copia por linha para manter cada geracao isolada
    return [linha[:] for linha in grade]


def criar_grade(linhas, colunas, percentual_espalhadores=0.02, semente=42):
    """Cria uma grade deterministica com espalhadores iniciais"""

    rng = random.Random(semente)
    grade = [[IGNORANTE for _ in range(colunas)] for _ in range(linhas)]

    total_celulas = linhas * colunas
    total_espalhadores = int(total_celulas * percentual_espalhadores)

    if total_espalhadores <= 0:
        return grade

    # A escolha sem repeticao garante a quantidade
    for posicao in rng.sample(range(total_celulas), total_espalhadores):
        i, j = divmod(posicao, colunas)
        grade[i][j] = ESPALHADOR

    return grade


def contar_vizinhos_espalhadores(grade, i, j):
    """Conta espalhadores na vizinhanca de Moore."""

    linhas = len(grade)
    colunas = len(grade[0])
    total = 0

    # Varrimento local dos 8 vizinhos ao redor da celula atual.
    for di in (-1, 0, 1):
        for dj in (-1, 0, 1):
            if di == 0 and dj == 0:
                continue

            ni = i + di
            nj = j + dj

            if 0 <= ni < linhas and 0 <= nj < colunas:
                if grade[ni][nj] == ESPALHADOR:
                    total += 1

    return total


def calcular_linha_proxima_geracao(grade, i, limiar_convencimento):
    """Calcula uma linha da proxima geracao a partir da grade atual."""

    nova_linha = grade[i][:]

    for j in range(len(grade[0])):
        estado_atual = grade[i][j]

        # A regra sempre usa a geracao anterior como referencia
        if estado_atual == IGNORANTE:
            vizinhos = contar_vizinhos_espalhadores(grade, i, j)
            nova_linha[j] = ESPALHADOR if vizinhos >= limiar_convencimento else IGNORANTE
        elif estado_atual == ESPALHADOR:
            nova_linha[j] = INATIVO
        else:
            nova_linha[j] = INATIVO

    return nova_linha


def proxima_geracao_sequencial(grade, limiar_convencimento=2):
    # A versao sequencial calcula linha por linha em uma nova matriz
    return [calcular_linha_proxima_geracao(grade, i, limiar_convencimento) for i in range(len(grade))]


def dividir_faixas(total_linhas, quantidade_partes):
    # Divide o trabalho em blocos adjacentes de linhas.
    if total_linhas <= 0:
        return []

    quantidade_partes = max(1, min(quantidade_partes, total_linhas))
    base = total_linhas // quantidade_partes
    resto = total_linhas % quantidade_partes

    faixas = []
    inicio = 0

    for indice in range(quantidade_partes):
        tamanho = base + (1 if indice < resto else 0)
        fim = inicio + tamanho
        if inicio < fim:
            faixas.append((inicio, fim))
        inicio = fim

    return faixas


def contar_estados(grade):
    # Retorna um resumo simples da populacao por estado.
    contagem = {
        IGNORANTE: 0,
        ESPALHADOR: 0,
        INATIVO: 0,
    }

    for linha in grade:
        for celula in linha:
            contagem[celula] += 1

    return contagem


def imprimir_grade(grade, limite=30):
    # Impressao parcial para nao poluir o terminal em grades maiores
    simbolos = {
        IGNORANTE: ".",
        ESPALHADOR: "E",
        INATIVO: "N",
    }

    linhas = min(len(grade), limite)
    colunas = min(len(grade[0]), limite)

    for i in range(linhas):
        print(" ".join(simbolos[grade[i][j]] for j in range(colunas)))
    print()


def construir_bloco_com_halos(grade, inicio, fim):
    """Monta um bloco com a linha anterior e a posterior como halo."""

    colunas = len(grade[0])
    bloco = []

    # O halo superior e inferior permitem calcular a vizinhanca nas bordas.
    if inicio > 0:
        bloco.append(grade[inicio - 1][:])
    else:
        bloco.append([IGNORANTE] * colunas)

    for linha in grade[inicio:fim]:
        bloco.append(linha[:])

    if fim < len(grade):
        bloco.append(grade[fim][:])
    else:
        bloco.append([IGNORANTE] * colunas)

    return bloco


def processar_bloco_distribuido(bloco, limiar_convencimento):
    # O worker processa apenas as linhas centrais do bloco recebido.
    if len(bloco) < 3:
        return []

    return [
        calcular_linha_proxima_geracao(bloco, i, limiar_convencimento)
        for i in range(1, len(bloco) - 1)
    ]
